### ISSUE TITLE
Should gluuGroup use groupOfEntries as its SUP objectClass?

### ISSUE THREAD

**Issue by madumlao** [GluuFederation/community-edition-setup#276]
The long and short of it is that OpenDJ (and I assume some other LDAP implementations) only recognizes some objectClasses as "LDAP groups" internally. With OpenDJ it is groupOfNames, groupOfUniqueNames, groupOfEntries, plus some other objectClasses that are treated as dynamic or virtual static groups.

For all intents and purposes, gluuGroups behave like "static groups" (groupOfEntries) with additional display / application logic. However, gluuGroups cannot be used in LDAP operations that check group membership, such as groupdn queries in ACIs, or virtual attributes, because gluuGroups are themselves not treated as one of the native LDAP group objectClasses.

Of the native LDAP group objects, groupOfEntries is the most appropriate for gluuGroups as the member attribute is optional, alllowing you to create empty groups. With groupOfNames or groupOfUniqueNames, member is required, so there must always be one group member. groupOfEntries is specified in the findlay draft (https://tools.ietf.org/html/draft-findlay-ldap-groupofentries-00) and was never officially RFC'd, but seems to have widespread use in directory servers.

I have found that it is trivial to "add" groupOfEntries to an existing gluuGroup entry, by adding a cn and objectClass groupOfEntries to the gluuGroup. This allows the group to be used in ACIs and other OpenDJ operations while at the same time allowing the group to be managed in oxTrust. Upon adding the objectClass everything seems to work as expected.

I feel that this should be the default, by setting the SUP objectClass of gluuGroup to groupOfEntries. The only issue I see with it is that the groupOfEntries CN and the gluuGroup displayName both seem to fulfill the same role and should be filled in with an oxTrust field. Or, CN might be a "unix" name and displayName be the descriptive name.

**Comment by nynymike** [comment]
Mark, I'm not a big fan of using groups for anything. What is your use case?

**Comment by madumlao** [comment]
You can probably guess that this my current use is for "legacy support".

The TLDR is that some apps dont have much of a choice outside of LDAP access (modifying their app or using middleware will break vendor support contracts) and I'm making allowances for them by creating limited "LDAP reader" accounts. For manageability, I created an ACI that allows members of a specific groupdn to have LDAP read access to gluuPerson and gluuGroup. A created gluuGroup, however, can't be used for this purpose (gluuGroups don't match a groupdn LDAP ACI), and an LDAP groupOfEntries, however, can't be managed in oxTrust (plus I dont want to open a whole can of worms with having the Gluu admin muck around unrestricted with the LDAP or worse, make user-specific LDAP acis). The middle ground is for me to force particular gluuGroups to also be an LDAP groupOfEntries with the ACI above, and allow the admin to manage membership in oxTrust.

For reference the ACI looks like this and works awesome, so long as the groupDN is an actual LDAP group.

~~~
(targetfilter="(|(objectClass=gluuPerson)(objectClass=gluuGroup))")(targetattr!="userPassword||authPassword||changes||changeNumber||changeType||changeTime||targetDN||newRDN||newSuperior||deleteOldRDN||targetEntryUUID||changeInitiatorsName||changeLogCookie||includedAttributes")(version 3.0; acl "LDAP user access"; allow (read,search,compare) groupdn="ldap:///GROUP_DN_HERE";)
~~~

Equivalent ACI rules can also be created in OpenLDAP, but only if the gluuGroup is an actual LDAP group.

---

But in the general case, I think it's useful for gluuGroups to also count as LDAP groups.

Probably the most important point is that it simply breaks user expectations. Groups feature prominently in the interface (it's how gluu manager permissions are set), LDAP features prominently in the documentation, people are going to try to use groups for something.  They were already previously used in 2.3 for client authorization and we've worked around legacy code depending on that behavior.

As for other uses:

App developers can rely on a member_of claim / memberOf attribute for custom logic for their app, for example, Zimbra users setting their "class of service" on import. Non-trivial example below:

User roles is the simplest case - simple pseudocode follows

~~~
if user has member_of $admins_group_dn:
    show_admin_page
else if user has member_of $doctors_group_dn:
    show_doctors_page
else if user has member_of $users_group_dn:
    show_regular_page
else:
    show_guest_page
end
~~~

Actually any arbitrary claim can be used for this purpose, which opens up apps to a ton of flexibility. Depending on how the user is imported / created (a custom registration page, an LDAP import script, bulk upload etc), gluuPersons can be preloaded with custom attributes. The presence or absence of such attributes can be used in the LDAP implementation to add virtual attributes, dynamic membership, group membership, etc - the end result simply being "more attributes for apps to use".

One of the ways virtual attributes (or their equivalent) can be used is based on group membership. So I might be able to put "zimbraCos: 1234" into a user based on membership in a "zimbraPremium" group. The alternative is to have any import and update scripts encode the attribute in the LDAP, but that's more flaky - requires more custom code, more points to test, etc. OTOH, the directory servers are already well-tested and we can rely on their group membership code.

**Comment by nynymike** [comment]
I think you can also make access policies based on user attributes. The "memberOf" attribute is pretty equivalent to using groups. It does not sound like you are using any attribute of group other then person. I'm closing this issue because I don't think it's a priority for our roadmap.
